kernel/src/utilities/arch_helpers.rs

```diff
@@ -0,0 +1,237 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Tock Contributors 2024.
+
+//! Helper functions and types shared between multiple `arch` crates.
+//!
+//! This function contains functions and types that do not have to be in the
+//! core kernel and are architecture-specific, but are shared by two or more
+//! `arch` crates. While these could also live in a dedicated crate, we use the
+//! `kernel` crate as all `arch` crates already depend on it.
+
+use crate::syscall::SyscallReturn;
+use crate::ErrorCode;
+
+/// Helper function to split a [`u64`] into a higher and lower [`u32`].
+///
+/// Used in encoding 64-bit wide system call return values on 32-bit
+/// platforms.
+#[inline]
+fn u64_to_be_u32s(src: u64) -> (u32, u32) {
+    let src_bytes = src.to_be_bytes();
+    let src_msb = u32::from_be_bytes([src_bytes[0], src_bytes[1], src_bytes[2], src_bytes[3]]);
+    let src_lsb = u32::from_be_bytes([src_bytes[4], src_bytes[5], src_bytes[6], src_bytes[7]]);
+
+    (src_msb, src_lsb)
+}
+
+/// Enumeration of the system call return type variant identifiers described in
+/// TRD104.
+///
+/// Each variant is associated with the respective variant identifier that would
+/// be passed along with the return value to userspace.
+#[repr(u32)]
+#[derive(Copy, Clone, Debug)]
+pub enum TRD104SyscallReturnVariant {
+    Failure = 0,
+    FailureU32 = 1,
+    FailureU32U32 = 2,
+    FailureU64 = 3,
+    Success = 128,
+    SuccessU32 = 129,
+    SuccessU32U32 = 130,
+    SuccessU64 = 131,
+    SuccessU32U32U32 = 132,
+    SuccessU32U64 = 133,
+}
+
+/// System call return variants defined as defined in TRD104. These are a strict
+/// subset of the variants defined in the core kernel's [`SyscallReturn`]
+/// enum. For documentation on the individual variants, refer to this type
+/// instead.
+#[derive(Copy, Clone, Debug)]
+pub enum TRD104SyscallReturn {
+    Failure(ErrorCode),
+    FailureU32(ErrorCode, u32),
+    FailureU32U32(ErrorCode, u32, u32),
+    FailureU64(ErrorCode, u64),
+    Success,
+    SuccessU32(u32),
+    SuccessU32U32(u32, u32),
+    SuccessU32U32U32(u32, u32, u32),
+    SuccessU64(u64),
+    SuccessU32U64(u32, u64),
+    AllowReadWriteSuccess(*mut u8, usize),
+    AllowReadWriteFailure(ErrorCode, *mut u8, usize),
+    UserspaceReadableAllowSuccess(*mut u8, usize),
+    UserspaceReadableAllowFailure(ErrorCode, *mut u8, usize),
+    AllowReadOnlySuccess(*const u8, usize),
+    AllowReadOnlyFailure(ErrorCode, *const u8, usize),
+    SubscribeSuccess(*const (), usize),
+    SubscribeFailure(ErrorCode, *const (), usize),
+    YieldWaitFor(usize, usize, usize),
```

Comment:
I don't think there should be any usize in this enum. I think everything expressing what is in the TRD104 should be explicitly u32. See the trd 104 refinement PR for rationale (sorry short text and no link on mobile and holding baby)

Reply:
#4228 for reference

I both have issues with the specificity of some of the types in #4228 and with pushing those down to here (even if `YieldWaitFor` is meant to return, e.g., all u32s) there isn't an issue with that conversion happening in `encode_syscall_return_trd104`

Reply:
Is this still something that we should address at this stage? This, right now, is just duplicating the `SyscallReturn` enum. I think that any changes to the individual variants would warrant their own PR, such as what #4174 will effectively be once this is merged.